tests/framework/utils.py

```diff
@@ -91,25 +92,73 @@ def get_cpu_percent(pid: int) -> Dict[str, Dict[str, float]]:
 class UffdHandler:
     """Describe the UFFD page fault handler process."""
 
-    def __init__(self, name, args):
+    def __init__(self, name, socket_path, mem_file, chroot, log_file_name):
         """Instantiate the handler process with arguments."""
         self._proc = None
-        self._args = [f"/{name}"]
-        self._args.extend(args)
+        self._handler_name = name
+        self._socket_path = socket_path
+        self._mem_file = mem_file
+        self._chroot = chroot
+        self._log_file = log_file_name
 
-    def spawn(self):
+    def spawn(self, uid, gid):
         """Spawn handler process using arguments provided."""
-        self._proc = subprocess.Popen(
-            self._args, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-        )
 
+        # Need to keep these around so we can exit the chroot
+        real_root = os.open("/", os.O_RDONLY)
+        working_dir = os.getcwd()
+
+        # Jump in the chroot
+        os.chroot(self._chroot)
+        os.chdir("/")
+        st = os.stat(self._handler_name)
+        os.chmod(self._handler_name, st.st_mode | stat.S_IEXEC)
+
+        chroot_log_file = Path("/") / self._log_file
+        with open(chroot_log_file, "w", encoding="utf-8") as logfile:
+            args = [f"/{self._handler_name}", self._socket_path, self._mem_file]
+            self._proc = subprocess.Popen(
```

Comment:
Could use preexec_fn to execute as uid/gid but not sure if it's worth the trouble: https://stackoverflow.com/questions/61302291/python-subprocess-popen-execute-as-different-user

Reply:
I tried that, but pylint complained that this function might not be safe in the presence of threads.